lib/action_policy/rspec/dsl.rb

```diff
@@ -78,10 +78,30 @@ def formatted_policy(policy)
       end
     else
       specify "is denied" do
-        next if subject.fail?
+        if subject.success?
+          raise(
+            RSpec::Expectations::ExpectationNotMetError,
+            "Expected to fail but succeed:\n#{formatted_policy(policy)}",
+            the_caller
+          )
+        end
+
+        case reason
+        when nil
+          next
+        when Hash
+          next if subject.reasons.details == reason
+        when Symbol
+          next if subject.reasons.details.values.flatten.any? { |h| h.include?(reason) }
+        when String
+          next if subject.reasons.respond_to?(:full_messages) && subject.reasons.full_messages.flatten.include?(reason)
+        else
+          raise TypeError, "unexpected reason: #{reason.inspect}"
+        end
```

Comment:
I would leave only the Hash matching option but add support for RSpec compound matchers instead:

```
failed "some desription", reason: a_hash_inclusing(no_admin: true)
```

For that, we only need to change the order and the `==` for `===`:

```ruby
reason === subject.reasons.details
```

Partial and especially nested matching seems a bit off; if we unit-test policies, we must take care of the actual _shape_.

I18n is also out of scope, IMO.